File: apps/mobile/src/lib/bootstrap.ts
import type { BootstrapState } from "../types/app";
import type { ApiBootstrapResponse } from "../types/api";

import { apiFetch, ApiError } from "./api";
import { clearAccessToken } from "./session";

function mapBootstrapResponse(response: ApiBootstrapResponse): BootstrapState {
  return {
    accountStatus: response.account_status,
    onboardingStep: response.onboarding_step,
    enabledTabs: response.enabled_tabs,
    supportedAuthProviders: response.supported_auth_providers,
    skippableOnboardingSteps: response.skippable_onboarding_steps,
  };
}

export async function getBootstrapState(): Promise<BootstrapState> {
  try {
    const response = await apiFetch<ApiBootstrapResponse>("/v1/bootstrap");
    return mapBootstrapResponse(response);
  } catch (error) {
    if (error instanceof ApiError && error.status === 401) {
      await clearAccessToken();
    }

    return {
      accountStatus: "unauthenticated",
      onboardingStep: null,
      enabledTabs: [],
      supportedAuthProviders: ["email", "phone", "google", "apple"],
      skippableOnboardingSteps: ["prompts", "interests", "lifestyle", "personality", "location"],
    };
  }
}

File: services/api/app/schemas/bootstrap.py
from enum import StrEnum

from pydantic import BaseModel


class AccountStatus(StrEnum):
    UNAUTHENTICATED = "unauthenticated"
    PENDING_VERIFICATION = "pending_verification"
    ONBOARDING_IN_PROGRESS = "onboarding_in_progress"
    ACTIVE = "active"
    PAUSED = "paused"
    BLOCKED = "blocked"
    DELETED = "deleted"


class AuthProvider(StrEnum):
    EMAIL = "email"
    PHONE = "phone"
    GOOGLE = "google"
    APPLE = "apple"


class OnboardingStepId(StrEnum):
    BASIC_PROFILE = "basic-profile"
    MEDIA = "media"
    PROMPTS = "prompts"
    INTERESTS = "interests"
    LIFESTYLE = "lifestyle"
    PERSONALITY = "personality"
    LOCATION = "location"
    PREFERENCES = "preferences"


class EnabledTab(StrEnum):
    PERSONALITY = "personality"
    LIKES = "likes"
    DISCOVER = "discover"
    CHAT = "chat"
    PROFILE = "profile"


class BootstrapResponse(BaseModel):
    user_id: str | None = None
    account_status: AccountStatus
    onboarding_step: OnboardingStepId | None = None
    is_profile_complete: bool
    enabled_tabs: list[EnabledTab]
    supported_auth_providers: list[AuthProvider]
    skippable_onboarding_steps: list[OnboardingStepId]
